File: src/main/java/com/demonwav/mcdev/creator/ForgeProjectSettingsWizard.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.creator;

import com.demonwav.mcdev.asset.PlatformAssets;
import com.demonwav.mcdev.platform.PlatformType;
import com.demonwav.mcdev.platform.forge.ForgeProjectConfiguration;
import com.demonwav.mcdev.platform.forge.version.ForgeVersion;
import com.demonwav.mcdev.platform.hybrid.SpongeForgeProjectConfiguration;
import com.demonwav.mcdev.platform.mcp.version.McpVersion;
import com.demonwav.mcdev.platform.mcp.version.McpVersionEntry;
import com.intellij.openapi.options.ConfigurationException;
import java.awt.event.ActionListener;
import java.util.List;
import javax.swing.JCheckBox;
import javax.swing.JComboBox;
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JProgressBar;
import javax.swing.JTextField;
import javax.swing.SwingWorker;
import org.apache.commons.lang.WordUtils;
import org.jetbrains.annotations.NotNull;

public class ForgeProjectSettingsWizard extends MinecraftModuleWizardStep {

    private JTextField pluginNameField;
    private JTextField pluginVersionField;
    private JTextField mainClassField;
    private JPanel panel;
    private JLabel title;
    private JTextField descriptionField;
    private JTextField authorsField;
    private JTextField websiteField;
    private JTextField dependField;
    private JTextField updateUrlField;
    private JComboBox<String> minecraftVersionBox;
    private JComboBox<String> forgeVersionBox;
    private JComboBox<McpVersionEntry> mcpVersionBox;
    private JProgressBar loadingBar;
    private JCheckBox generateDocsCheckbox;
    private JLabel minecraftVersionLabel;
    private JLabel mcpWarning;

    private ForgeProjectConfiguration settings;
    private final MinecraftProjectCreator creator;

    private McpVersion mcpVersion;
    private ForgeVersion forgeVersion;

    private final ActionListener mcpBoxActionListener = e -> {
        if (((McpVersionEntry) mcpVersionBox.getSelectedItem()).isRed()) {
            mcpWarning.setVisible(true);
        } else {
            mcpWarning.setVisible(false);
        }
    };

    private ForgeWorker apiWorker = new ForgeWorker();

    public ForgeProjectSettingsWizard(@NotNull MinecraftProjectCreator creator) {
        this.creator = creator;

        generateDocsCheckbox.setVisible(false);
        mcpWarning.setVisible(false);

        minecraftVersionBox.addActionListener(e -> {
            if (mcpVersion != null) {
                mcpVersion.setMcpVersion(mcpVersionBox, getVersion(), mcpBoxActionListener);
            }
            setForgeVersion();
        });
    }

    @Override
    public JComponent getComponent() {
        settings = (ForgeProjectConfiguration) creator.getSettings().get(PlatformType.FORGE);
        if (settings == null) {
            return null;
        }

        pluginNameField.setText(WordUtils.capitalize(creator.getArtifactId()));
        pluginVersionField.setText(creator.getVersion());

        if (settings != null && !settings.isFirst) {
            pluginNameField.setEditable(false);
            pluginVersionField.setEditable(false);
        }

        mainClassField.setText(this.creator.getGroupId().toLowerCase() + '.' + this.creator.getArtifactId().toLowerCase()
                + '.' + WordUtils.capitalize(this.creator.getArtifactId()));

        if (creator.getSettings().size() > 1) {
            mainClassField.setText(mainClassField.getText() + PlatformType.FORGE.getNormalName());
        }

        loadingBar.setIndeterminate(true);

        if (settings instanceof SpongeForgeProjectConfiguration) {
            title.setIcon(PlatformAssets.SPONGE_FORGE_ICON_2X);
            title.setText("<html><font size=\"5\">SpongeForge Settings</font></html>");
            generateDocsCheckbox.setVisible(true);

            minecraftVersionLabel.setText("    Sponge API");
        }

        apiWorker.execute();

        return panel;
    }

    @Override
    public void updateStep() {
        if ((forgeVersion == null || mcpVersion == null) && (apiWorker.isCancelled() || apiWorker.isDone())) {
            // A SwingWorker will only run once, so we need to create a new instance
            apiWorker = new ForgeWorker();
            apiWorker.execute();
        }
    }

    private void setForgeVersion() {
        if (forgeVersion == null) {
            return;
        }

        String version = getVersion();

        if (version == null) {
            return;
        }

        forgeVersionBox.removeAllItems();
        List<String> versions = forgeVersion.getForgeVersions(version);

        if (versions == null) {
            return;
        }
        versions.stream().sorted((one, two) -> one.compareTo(two) * -1).forEach(forgeVersionBox::addItem);

        Double promo = forgeVersion.getPromo(version);
        if (promo != null) {
            int index = 0;
            for (int i = 0; i < forgeVersionBox.getItemCount(); i++) {
                try {
                    if (forgeVersionBox.getItemAt(i).endsWith(String.valueOf(promo.intValue()))) {
                        index = i;
                    }
                } catch (NumberFormatException ignored) {}
            }
            forgeVersionBox.setSelectedIndex(index);
        }
    }

    private String getVersion() {
        String version;
        if (!(settings instanceof SpongeForgeProjectConfiguration)) {
            version = (String) minecraftVersionBox.getSelectedItem();
        } else {
            if (minecraftVersionBox.getSelectedItem().equals("4.1.0")) {
                version = "1.8.9";
            } else {
                version = "1.10.2";
            }
        }
        return version;
    }

    @Override
    public boolean validate() throws ConfigurationException {
        return validate(pluginNameField, pluginVersionField, mainClassField, authorsField, dependField, pattern) && !loadingBar.isVisible();
    }

    @Override
    public boolean isStepVisible() {
        settings = (ForgeProjectConfiguration) creator.getSettings().get(PlatformType.FORGE);
        return settings != null;
    }

    @Override
    public void onStepLeaving() {
        if (loadingBar.isVisible()) {
            // we're in a cancel state
            apiWorker.cancel(true);
            return;
        }

        settings.pluginName = pluginNameField.getText();
        settings.pluginVersion = pluginVersionField.getText();
        settings.mainClass = mainClassField.getText();

        settings.setAuthors(authorsField.getText());
        settings.setDependencies(dependField.getText());
        settings.description = descriptionField.getText();
        settings.website = websiteField.getText();
        settings.updateUrl = updateUrlField.getText();

        settings.mcpVersion = ((McpVersionEntry) mcpVersionBox.getSelectedItem()).getText();

        if (settings instanceof SpongeForgeProjectConfiguration) {
            SpongeForgeProjectConfiguration configuration = (SpongeForgeProjectConfiguration) settings;
            configuration.generateDocumentation = generateDocsCheckbox.isSelected();
            configuration.spongeApiVersion = (String) minecraftVersionBox.getSelectedItem();
        }

        // If an error occurs while fetching the API, this may prevent the user from closing the dialog.
        try {
            settings.forgeVersion = forgeVersion.getFullVersion((String) forgeVersionBox.getSelectedItem());
        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void updateDataModel() {}

    private class ForgeWorker extends SwingWorker<Object, Object> {

        @Override
        protected Object doInBackground() throws Exception {
            mcpVersion = McpVersion.downloadData();
            forgeVersion = ForgeVersion.downloadData();
            return null;
        }

        @Override
        protected void done() {
            if (mcpVersion == null) {
                return;
            }

            minecraftVersionBox.removeAllItems();

            // reverse order the versions
            if (!(settings instanceof SpongeForgeProjectConfiguration)) {
                forgeVersion.getSortedMcVersions().forEach(minecraftVersionBox::addItem);
                final String recommended = forgeVersion.getRecommended(mcpVersion.getVersions());

                int index = 0;
                for (int i = 0; i < minecraftVersionBox.getItemCount(); i++) {
                    if (minecraftVersionBox.getItemAt(i).equals(recommended)) {
                        index = i;
                    }
                }
                minecraftVersionBox.setSelectedIndex(index);
            } else {
                minecraftVersionBox.addItem("4.1.0");
                minecraftVersionBox.addItem("5.0.0");
                minecraftVersionBox.setSelectedIndex(1);
            }

            if (mcpVersion != null) {
                mcpVersion.setMcpVersion(mcpVersionBox, getVersion(), mcpBoxActionListener);
            }

            if (forgeVersion == null) {
                return;
            }

            setForgeVersion();

            loadingBar.setIndeterminate(false);
            loadingBar.setVisible(false);
        }
    }
}

File: src/main/kotlin/com/demonwav/mcdev/creator/ForgeProjectSettingsWizard.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.creator

import com.demonwav.mcdev.asset.PlatformAssets
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.ProjectConfiguration
import com.demonwav.mcdev.platform.forge.ForgeProjectConfiguration
import com.demonwav.mcdev.platform.forge.version.ForgeVersion
import com.demonwav.mcdev.platform.mcp.version.McpVersion
import com.demonwav.mcdev.platform.mcp.version.McpVersionEntry
import com.demonwav.mcdev.util.firstOfType
import com.intellij.ui.CollectionComboBoxModel
import java.awt.event.ActionListener
import javax.swing.JComboBox
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JProgressBar
import javax.swing.JTextField
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import org.apache.commons.lang.WordUtils

class ForgeProjectSettingsWizard(private val creator: MinecraftProjectCreator) : MinecraftModuleWizardStep() {

    private lateinit var modNameField: JTextField
    private lateinit var modVersionField: JTextField
    private lateinit var mainClassField: JTextField
    private lateinit var panel: JPanel
    private lateinit var title: JLabel
    private lateinit var descriptionField: JTextField
    private lateinit var authorsField: JTextField
    private lateinit var websiteField: JTextField
    private lateinit var updateUrlField: JTextField
    private lateinit var minecraftVersionBox: JComboBox<String>
    private lateinit var forgeVersionBox: JComboBox<String>
    private lateinit var mcpVersionBox: JComboBox<McpVersionEntry>
    private lateinit var loadingBar: JProgressBar
    private lateinit var minecraftVersionLabel: JLabel
    private lateinit var mcpWarning: JLabel
    private lateinit var errorLabel: JLabel

    private var config: ForgeProjectConfiguration? = null

    private data class ForgeVersions(
        var mcpVersion: McpVersion,
        var forgeVersion: ForgeVersion
    )

    private var versions: ForgeVersions? = null

    private var currentJob: Job? = null

    private val mcpBoxActionListener = ActionListener {
        mcpWarning.isVisible = (mcpVersionBox.selectedItem as? McpVersionEntry)?.isRed == true
    }

    private val minecraftBoxActionListener: ActionListener = ActionListener {
        CoroutineScope(Dispatchers.Swing).launch {
            loadingBar.isIndeterminate = true
            loadingBar.isVisible = true

            updateForm()

            loadingBar.isIndeterminate = false
            loadingBar.isVisible = false
        }
    }

    init {
        mcpWarning.isVisible = false
    }

    override fun getComponent(): JComponent {
        return panel
    }

    override fun updateStep() {
        config = creator.configs.firstOfType()

        val buildSystem = creator.buildSystem ?: return

        modNameField.text = WordUtils.capitalize(buildSystem.artifactId.replace('-', ' '))
        modVersionField.text = buildSystem.version

        val conf = config ?: return

        if (creator.configs.indexOf(conf) != 0) {
            modNameField.isEditable = false
            modVersionField.isEditable = false
        }

        mainClassField.text = buildSystem.groupId.replace("-", "").toLowerCase() + "." +
            buildSystem.artifactId.replace("-", "").toLowerCase() + "." +
            WordUtils.capitalize(buildSystem.artifactId.replace('-', ' ')).replace(" ", "")

        if (creator.configs.size > 1) {
            mainClassField.text = mainClassField.text + PlatformType.FORGE.normalName
        }

        title.icon = PlatformAssets.FORGE_ICON_2X
        title.text = "<html><font size=\"5\">Forge Settings</font></html>"

        minecraftVersionLabel.text = "Minecraft Version"

        if (versions != null || currentJob?.isActive == true) {
            return
        }
        currentJob = updateVersions()
    }

    private fun setForgeVersion(data: Data) {
        forgeVersionBox.model = CollectionComboBoxModel(data.forgeVersions)
        forgeVersionBox.selectedIndex = data.forgeSelectedIndex
    }

    private val version: String?
        get() = minecraftVersionBox.selectedItem as? String

    override fun validate(): Boolean {
        return validate(
            modNameField,
            modVersionField,
            mainClassField,
            authorsField,
            null,
            pattern
        ) && !loadingBar.isVisible
    }

    override fun isStepVisible(): Boolean {
        return creator.configs.any { it is ForgeProjectConfiguration }
    }

    override fun onStepLeaving() {
        currentJob?.let { job ->
            // we're in a cancel state
            job.cancel()
            return
        }

        val conf = config ?: return
        conf.base = ProjectConfiguration.BaseConfigs(
            pluginName = modNameField.text,
            pluginVersion = modVersionField.text,
            mainClass = mainClassField.text,
            description = descriptionField.text,
            website = websiteField.text
        )

        conf.setAuthors(authorsField.text)
        conf.updateUrl = updateUrlField.text

        conf.mcpVersion = (mcpVersionBox.selectedItem as McpVersionEntry).versionPair

        (forgeVersionBox.selectedItem as? String)?.let { version ->
            val forgeVersion = versions?.forgeVersion ?: return@let
            conf.forgeVersion = forgeVersion.versions.first { it.endsWith(version) }
        }

        conf.mcVersion = version ?: ""
    }

    fun error() {
        errorLabel.isVisible = true
        loadingBar.isIndeterminate = false
        loadingBar.isVisible = false
    }

    override fun updateDataModel() {}

    private fun mcVersionUpdate(data: Data) {
        mcpVersionBox.removeActionListener(mcpBoxActionListener)
        mcpVersionBox.model = CollectionComboBoxModel(data.mcpVersions)
        mcpVersionBox.selectedIndex = 0
        mcpVersionBox.addActionListener(mcpBoxActionListener)
        mcpBoxActionListener.actionPerformed(null)

        setForgeVersion(data)
    }

    private fun updateVersions() = CoroutineScope(Dispatchers.Swing).launch {
        loadingBar.isIndeterminate = true
        loadingBar.isVisible = true

        try {
            downloadVersions()
            val data = updateForm()
            if (data != null) {
                updateMcForm(data)
            }
        } catch (e: Exception) {
            error()
        }

        loadingBar.isIndeterminate = false
        loadingBar.isVisible = false

        currentJob = null
    }

    private suspend fun downloadVersions() = coroutineScope {
        val mcpVersionJob = async(Dispatchers.IO) { McpVersion.downloadData() }
        val forgeVersionJob = async(Dispatchers.IO) { ForgeVersion.downloadData() }

        versions = ForgeVersions(
            mcpVersionJob.await() ?: return@coroutineScope,
            forgeVersionJob.await() ?: return@coroutineScope
        )
    }

    private suspend fun updateForm(): Data? = coroutineScope {
        val vers = versions ?: return@coroutineScope null

        val finalVersion = when {
            version != null -> {
                // selected version
                version ?: return@coroutineScope null
            }
            else -> {
                // Default Forge
                vers.forgeVersion.sortedMcVersions.firstOrNull() ?: return@coroutineScope null
            }
        }

        val mcpVersionListJob = async(Dispatchers.IO) { vers.mcpVersion.getMcpVersionList(finalVersion) }
        val forgeVersionsJob = async(Dispatchers.IO) {
            val list = vers.forgeVersion.getForgeVersions(finalVersion)
            list.sortDescending()
            for (i in 0 until list.size) {
                list[i] = list[i].substring(list[i].indexOf('-') + 1)
            }
            return@async list
        }

        val mcpVersionList = mcpVersionListJob.await()
        val forgeVersions = forgeVersionsJob.await()

        val forgeIndex = 0

        val data = Data(0, mcpVersionList, forgeVersions, forgeIndex)

        mcVersionUpdate(data)

        return@coroutineScope data
    }

    private fun updateMcForm(data: Data) {
        val vers = versions ?: return

        minecraftVersionBox.removeActionListener(minecraftBoxActionListener)
        minecraftVersionBox.removeAllItems()

        minecraftVersionBox.model = CollectionComboBoxModel(vers.forgeVersion.sortedMcVersions)
        minecraftVersionBox.selectedIndex = data.mcSelectedIndex
        minecraftVersionBox.addActionListener(minecraftBoxActionListener)
    }

    private data class Data(
        val mcSelectedIndex: Int,
        val mcpVersions: List<McpVersionEntry>,
        val forgeVersions: List<String>,
        val forgeSelectedIndex: Int
    )
}
